Added new stacking option, category-center

TorsteinHonsi opened the pull request:
Added new option,  `plotOptions.column.centerInCategory`. It centers columns within each category group, but avoiding gaps for `null` or missing points.

Fixes #11308.

This PR offers an alternative to #12818. By tapping into the existing stacking logic, we avoid having to create lots of logic that basically deals with the same problem - points on the same X value need to be aware of each other. By using the stacking logic we simply reduce the logic to compute the final position of the columns.

### To do
 - [x] Update sample and docs
 - [x] Implement in combination with simple and multiple stacks
 - [x] Check out multiple points at the same X value
 - [x] Move over and adapt relevant unit tests from #12818
 - [x] Fix two issues mentioned in [comment](https://github.com/highcharts/highcharts/pull/12846#issuecomment-579196086) (Torstein).

circleci-mu commented:
## Visual test results - No difference found

---
<details>
<summary>Samples changed</summary><p>

| Change type | Sample |
| --- | --- |
|  Added | samples/highcharts/plotoptions/series-stacking-category-center/demo.js |

</p>
</details>

pawelfus commented:
Nice! 
But I have one problem with this approach: user can't use regular stacking with stacks **and** centered columns, for example this output is impossible (manually edited SVG):
- (`stack: 'A', stack: 'B'` + `stacking: 'normal'` + centered columns):
![Screen Shot 2020-01-24 at 13 03 53](https://user-images.githubusercontent.com/1453926/73068076-9a299400-3eaa-11ea-915c-87a15f7b0bf0.png)

We can get only one of those:
- (`stack: 'A', stack: 'B'` + `stacking: 'normal'`):
![Screen Shot 2020-01-24 at 13 05 53](https://user-images.githubusercontent.com/1453926/73068077-9a299400-3eaa-11ea-904f-b78156809729.png)

- (`stack: 'A', stack: 'B'` + `stacking: 'category-center'`):
![Screen Shot 2020-01-24 at 13 06 14](https://user-images.githubusercontent.com/1453926/73068078-9a299400-3eaa-11ea-831d-55dcf66f6379.png)

TorsteinHonsi commented:
> user can't use regular stacking with stacks and centered columns

Yes... the question is if that edge case is worth all the extra code. Alternatively, we should be able to combine the two concepts to avoid parallel code.

pawelfus commented:
I'm not sure if we can call it an edge case: users simply want columns centered over a category. Just a design change, position of a column.

Stacking in this case is just visual representation of a bar components. For me it's another feature that I would expect to be compatible with the solution above.

raf18seb commented:
Nice, I agree that the performance is much better and we have less new code. But, as Paweł said, we are missing stacking + centering and also we are missing more customization options like e.g. "fill available space" (columns have different width depending on a number of points in category).

Also, I got used to filling the whole available space by setting `series.pointPadding` and `series.groupPadding` to 0: http://jsfiddle.net/BlackLabel/y5za097e/
There's also a small problem with adding multiple same points.

TorsteinHonsi commented:
Okay, see the new commit. I now removed the centering from the `stacking` options, to avoid conflicts there. So now it should work in combination with `stacking` (except it looks like crap currently :).

The new option is `centerInCategory`. 

The change is that is still uses the stacking logic, but instead of being a separate stack option, it creates its own stack just for this purpose. This is not a new concept, Highcharts already supports positive and negative stacks for the same series/points, so I just added a `group` stack.

With this approach we get all the X-value mapping and dynamic stuff for free.

Now we need to figure out how _real_ stacks affects the positioning. 

> we are missing more customization options like e.g. "fill available space"

If we want those, it should be simple to implement it in the last scaling step in `ColumnSeries.translate`.

pawelfus commented on ts/parts/Stacking.ts:
```diff
@@ -661,8 +670,14 @@ Axis.prototype.cleanStacks = function (this: Highcharts.Axis): void {
-    if (!this.options.stacking ||
+Series.prototype.setStackedPoints = function (
+    this: Highcharts.Series,
+    stackingParam: string
```
I think this param should be optional

pawelfus commented on ts/parts/Stacking.ts:
```diff
@@ -652,7 +656,12 @@ Axis.prototype.cleanStacks = function (this: Highcharts.Axis): void {
+// Stacking methods defined for Series prototype
+Series.prototype.setGroupedPoints = function (this: Highcharts.Series): void {
+    if (this.options.centerInCategory) {
+        Series.prototype.setStackedPoints.call(this, 'group');
```
Hmm.. why not simply `this.setStackedPoints('group')`?

TorsteinHonsi replied:
Because many of the box-like series types override `setStackedPoints`. Haven't tested it though.

pawelfus replied:
Ah, that's the case! Thanks 👍

TorsteinHonsi commented:
I think I've got it working with stacking now. Tested no stacking, single stack and multiple stacks. @raf18seb will you move over and adapt the relevant unit tests?

raf18seb commented:
Still isn't working with the different series type (see commented line series) and isn't bulletproof for data commented in line 19: http://jsfiddle.net/BlackLabel/9gk152zv/

circleci-mu commented:
### File size comparison
| | master | candidate | difference |
|-------------|-------------:|-------------:|-------------:|
| highcharts | 262.0 kB | 262.6 kB | 700 B |
| highcharts, gzipped | 90.7 kB | 91.0 kB | 290 B|
| highstock | 340.3 kB | 341.0 kB | 704 B |
| highstock, gzipped | 116.7 kB | 117.0 kB | 325 B|
| highmaps | 310.4 kB | 311.0 kB | 700 B |
| highmaps, gzipped | 106.1 kB | 106.4 kB | 280 B|
| highcharts-gantt | 352.7 kB | 353.4 kB | 700 B |
| highcharts-gantt, gzipped | 120.8 kB | 121.1 kB | 308 B|

sebastianbochan commented:
Im curious, what is the reason of removing / comment test? Is it temporary solution for Safari?

shashank1010 commented:
Just curious to know by when would this be tagged in a release.

sebastianbochan commented:
The release should be available this week (if all test will pass).
